### ISSUE TITLE
Keypoints returned from callback not in the same position with AnnotationOverlay render

### ISSUE THREAD

**Issue by sufiaji** [google/mediapipe#654]
Hi Team,

First I need to say thank you for your great work! 
I managed to retrieved face keypoints from FrameProcessor callback. But when I draw the keypoints, the position is not the same with what mediapipe draws in AnnotationOverlayCalculator. As I understand, these keypoint coordinates are normalized coordinates, so I convert to pixel coordinate before drawing to the screen, with below code:

`// get 1st object of detection`
`DetectionProto.Detection det = detections.get(0);`
`// get keypoints`
`List<LocationDataProto.LocationData.RelativeKeypoint> rks = det.getLocationData().getRelativeKeypointsList();`
`// get width and height of SurfaceView`
`int W = previewDisplayView.getWidth();`
`int H = previewDisplayView.getHeight();`
`// create a new bitmap with same height/width of SurfaceView`
`final Bitmap bm = Bitmap.createBitmap(W, H, Bitmap.Config.ARGB_8888);`
`final HandlerThread handlerThread = new HandlerThread("PixelCopier");`
`handlerThread.start();`
`// copy SurfaceView image to the new bitmap`
`PixelCopy.request(previewDisplayView, /* rect, */ bm, (copyResult) -> {`
`    if (copyResult == PixelCopy.SUCCESS) {`
`        Log.e(TAG, bm.toString());`
`        // draw keypoints to new canvas`
`        Bitmap bmOverlay = Bitmap.createBitmap(bm.getWidth(), bm.getHeight(), bm.getConfig());`
`        Canvas canvas = new Canvas(bmOverlay);`
`        Paint paint = new Paint();`
`        canvas.drawBitmap(bm, new Matrix(), null);`
`        // draw the keypoints as blue small circle`
`        for (LocationDataProto.LocationData.RelativeKeypoint rk : rks) {`
`            String label = rk.getKeypointLabel();`
`            float x = rk.getX();`
`           float y = rk.getY();`
`            String msg = label + ": "+ x +", " + y;`
`           Log.d(TAG, msg);`
`            canvas.drawCircle(W*x, H*y, 10, paint);   }`
` ........... `

So basically I "re-normalized" it against the width/height of SurfaceView. Is it correct?
The keypoints are bit skewed (see image below, the blue dots are callback keypoints, the red dots are what mediapipe draws)

![keypoints_issue](https://user-images.githubusercontent.com/7789546/80299115-0a9ec480-87bc-11ea-93ce-9cb12414a2dc.jpg)

And other things, I got empty keypoints label also from callback.

![labels](https://user-images.githubusercontent.com/7789546/80299172-7f71fe80-87bc-11ea-811c-e7ae80f8744e.jpg)

Any clues on this problem?

Thanks!

**Comment by sufiaji** [comment]
Hi Team,

This is my callback code to get detections list

`processor.addPacketCallback(`
`output_detections,`
`List<DetectionProto.Detection> detections = PacketGetter.getProtoVector(packet, DetectionProto.Detection.parser());`

Thanks!

**Comment by sufiaji** [comment]
Hi Team,

The face mesh also got the same case. 

![photo_2020-04-27_15-52-39](https://user-images.githubusercontent.com/7789546/80353708-eddbbd00-889f-11ea-97cb-c6ad19b70646.jpg)

Do you have any suggestion @jiuqiant ?

Thanks in advance.

**Comment by sufiaji** [comment]
Ok I solved the problem by scaling the canvas before plotting the coordinate. I scaled the canvas by Wa/Wb; where Wa is the previewDisplayView width (screen width) and Wb is the previewFrameTexture width (640px). So instead de-normalizing only, we also need to scale. The weird thing is, I just need to scale X coordinate and no need to scale Y coordinate. However the size of previewFrameTexture  is 640x640...

**Comment by sufiaji** [comment]
I close this issue since the coordinate problem solved. Thanks team.

**Comment by shadow111** [comment]
> Hi Team,
> 
> First I need to say thank you for your great work!
> I managed to retrieved face keypoints from FrameProcessor callback. But when I draw the keypoints, the position is not the same with what mediapipe draws in AnnotationOverlayCalculator. As I understand, these keypoint coordinates are normalized coordinates, so I convert to pixel coordinate before drawing to the screen, with below code:
> 
> `// get 1st object of detection`
> `DetectionProto.Detection det = detections.get(0);`
> `// get keypoints`
> `List<LocationDataProto.LocationData.RelativeKeypoint> rks = det.getLocationData().getRelativeKeypointsList();`
> `// get width and height of SurfaceView`
> `int W = previewDisplayView.getWidth();`
> `int H = previewDisplayView.getHeight();`
> `// create a new bitmap with same height/width of SurfaceView`
> `final Bitmap bm = Bitmap.createBitmap(W, H, Bitmap.Config.ARGB_8888);`
> `final HandlerThread handlerThread = new HandlerThread("PixelCopier");`
> `handlerThread.start();`
> `// copy SurfaceView image to the new bitmap`
> `PixelCopy.request(previewDisplayView, /* rect, */ bm, (copyResult) -> {`
> ` if (copyResult == PixelCopy.SUCCESS) {`
> ` Log.e(TAG, bm.toString());`
> ` // draw keypoints to new canvas`
> ` Bitmap bmOverlay = Bitmap.createBitmap(bm.getWidth(), bm.getHeight(), bm.getConfig());`
> ` Canvas canvas = new Canvas(bmOverlay);`
> ` Paint paint = new Paint();`
> ` canvas.drawBitmap(bm, new Matrix(), null);`
> ` // draw the keypoints as blue small circle`
> ` for (LocationDataProto.LocationData.RelativeKeypoint rk : rks) {`
> ` String label = rk.getKeypointLabel();`
> ` float x = rk.getX();`
> ` float y = rk.getY();`
> ` String msg = label + ": "+ x +", " + y;`
> ` Log.d(TAG, msg);`
> ` canvas.drawCircle(W*x, H*y, 10, paint); }`
> `...........`
> 
> So basically I "re-normalized" it against the width/height of SurfaceView. Is it correct?
> The keypoints are bit skewed (see image below, the blue dots are callback keypoints, the red dots are what mediapipe draws)
> 
> ![keypoints_issue](https://user-images.githubusercontent.com/7789546/80299115-0a9ec480-87bc-11ea-93ce-9cb12414a2dc.jpg)
> 
> And other things, I got empty keypoints label also from callback.
> 
> ![labels](https://user-images.githubusercontent.com/7789546/80299172-7f71fe80-87bc-11ea-811c-e7ae80f8744e.jpg)
> 
> Any clues on this problem?
> 
> Thanks!

can you please share the code on how you draw the key points using android Paint as well as the logic behind. i have bloked the whole week trying to do the same things.
Thank you in advance
